### DISCUSSION TITLE
How do I calculate the bandwidth of the PCIe sockets? Why are only 2 sockets listed?

### DISCUSSION THREAD

**Discussion by KevinLucidyne** [opcm/pcm#127] [Q&A] [answered]
I'm working in deep learning and I'm trying to identify a bottleneck in our GPU pipeline.

We're running Ubuntu on an Intel Xeon motherboard with 4 NVIDIA Titan RTXs.  Use of the GPUs seems to be pretty low even though the memory usage is around 97%.

So I'm trying to see if the bus is the bottleneck.

I've downloaded [PCM][1] and I'm running it to monitor the PCIe 3.0 x16 traffic.

    Processor Counter Monitor: PCIe Bandwidth Monitoring Utility 
     This utility measures PCIe bandwidth in real-time
    
     PCIe event definitions (each event counts as a transfer): 
       PCIe read events (PCI devices reading from memory - application writes to disk/network/PCIe device):
         PCIeRdCur* - PCIe read current transfer (full cache line)
             On Haswell Server PCIeRdCur counts both full/partial cache lines
         RFO*      - Demand Data RFO
         CRd*      - Demand Code Read
         DRd       - Demand Data Read
       PCIe write events (PCI devices writing to memory - application reads from disk/network/PCIe device):
         ItoM      - PCIe write full cache line
         RFO       - PCIe partial Write
       CPU MMIO events (CPU reading/writing to PCIe devices):
         PRd       - MMIO Read [Haswell Server only] (Partial Cache Line)
         WiL       - MMIO Write (Full/Partial)
    ...
    Socket 0: 2 memory controllers detected with total number of 6 channels. 3 QPI ports detected. 2 M2M (mesh to memory) blocks detected.
    Socket 1: 2 memory controllers detected with total number of 6 channels. 3 QPI ports detected. 2 M2M (mesh to memory) blocks detected.
    Trying to use Linux perf events...
    Successfully programmed on-core PMU using Linux perf
    Link 3 is disabled
    Link 3 is disabled
    Socket 0
    Max QPI link 0 speed: 23.3 GBytes/second (10.4 GT/second)
    Max QPI link 1 speed: 23.3 GBytes/second (10.4 GT/second)
    Socket 1
    Max QPI link 0 speed: 23.3 GBytes/second (10.4 GT/second)
    Max QPI link 1 speed: 23.3 GBytes/second (10.4 GT/second)
    
    Detected Intel(R) Xeon(R) Gold 5122 CPU @ 3.60GHz "Intel(r) microarchitecture codename Skylake-SP" stepping 4 microcode level 0x200004d
    Update every 1.0 seconds
    delay_ms: 54
    Skt | PCIeRdCur |  RFO  |  CRd  |  DRd  |  ItoM  |  PRd  |  WiL
     0      13 K        19 K     0       0      220 K    84     588  
     1       0        3024       0       0        0       0     264  
    -----------------------------------------------------------------------
     *      13 K        22 K     0       0      220 K    84     852  

Ignore the actual values for a moment.  I have much bigger values. :-)

How do I calculate the bandwidth of the PCIe socket?

Why are there only two sockets listed?

  [1]: https://github.com/opcm/pcm

**Comment by opcm** [comment, marked as answer]
socket refer to CPU sockets not PCIe slots or devices. You have a system with 2 CPU sockets, right?
The --help output describes -B switch to output bandwidth per CPU socket (see also caveat here with partial operations). pci-iio.x is a different utility that shows bandwidth per PCIe device.
